contracts/core/src/token/MerkleDistributorWithDeadline.sol

```diff
@@ -0,0 +1,129 @@
+     * @param merkleProof  Merkle proof for the claim
+     */
+    function claim(uint256 index, address account, uint256 amount, bytes32[] calldata merkleProof) public override {
+        if (block.timestamp > endTime) revert ClaimWindowFinished();
+        super.claim(index, account, amount, merkleProof);
+    }
+
+    /**
+     * @notice Claim rewards and/or migrate stake to Omni
+     * @dev Triggers a SolverNet order to generate a subsidized order for deposited tokens on Omni 1:1
+     *      If the user has already claimed rewards, they can still migrate their stake to Omni
+     * @param index        Index of the claim
+     * @param amount       Amount of tokens to claim
+     * @param merkleProof  Merkle proof for the claim
+     */
+    function migrateToOmni(uint256 index, uint256 amount, bytes32[] calldata merkleProof) external {
+        if (block.timestamp > endTime) revert ClaimWindowFinished();
+
+        // Verify the merkle proof.
+        bytes32 node = keccak256(abi.encodePacked(index, msg.sender, amount));
+        if (!MerkleProofLib.verifyCalldata(merkleProof, merkleRoot, node)) revert InvalidProof();
+
+        // Migrate user's stake to Omni, if any
+        uint256 stake = IGenesisStake(genesisStaking).migrateStake(msg.sender);
+
+        // Mark reward distribution as claimed and add it to user's stake
+        if (!isClaimed(index)) {
+            claimedBitMap.set(index);
+            unchecked {
+                stake += amount;
+            }
+        }
+
+        // Generate and send the order
+        IERC7683.OnchainCrossChainOrder memory order = _generateOrder(msg.sender, stake);
+        IOriginSettler(solverNetInbox).open(order);
+    }
+
+    function withdraw() external onlyOwner {
+        if (block.timestamp < endTime) revert NoWithdrawDuringClaim();
+        token.safeTransfer(msg.sender, token.balanceOf(address(this)));
+    }
+
+    /**
+     * @notice Generate a SolverNet order that generates a subsidized order for deposited tokens on Omni 1:1
+     * @param account  Address of the user claiming
+     * @param amount   Amount of tokens to claim
+     * @return         SolverNet order
+     */
+    function _generateOrder(address account, uint256 amount)
+        internal
+        view
+        returns (IERC7683.OnchainCrossChainOrder memory)
+    {
+        ISolverNet.Deposit[] memory deposits = new ISolverNet.Deposit[](1);
+        deposits[0] = ISolverNet.Deposit({ token: token.toBytes32(), amount: amount });
+
+        ISolverNet.TokenExpense[] memory expenses = new ISolverNet.TokenExpense[](1);
+        expenses[0] =
+            ISolverNet.TokenExpense({ token: address(0).toBytes32(), spender: address(0).toBytes32(), amount: amount });
+
+        ISolverNet.Call memory call = ISolverNet.Call({
+            chainId: OMNI_CHAINID,
+            target: account.toBytes32(),
+            value: amount,
+            data: "",
+            expenses: expenses
+        });
```

Comment:
This is bridges stake to native omni.

I imaging we might want single click migration to native omni staking (in Staking.sol)

That would require letting users choose a validator to delegate too. Given we have allow list of validators, this is okay for now, as that set won't change. 

Check with Fab on UX here.